src/providers/powervs/mod.rs

```diff
@@ -0,0 +1,392 @@
+    /// Instance ID (UUID).
+    #[serde(rename = "uuid")]
+    pub instance_id: String,
+    /// SSH public keys.
+    pub public_keys: Option<HashMap<String, String>>,
+}
+
+/// Partial object for `network_data.json`
+#[derive(Debug, Deserialize)]
+pub struct NetworkDataJSON {
+    pub links: Vec<NetLinkJSON>,
+    pub networks: Vec<NetNetworkJSON>,
+    pub services: Vec<NetServiceJSON>,
+}
+
+/// JSON entry in `links` array.
+#[derive(Debug, Deserialize)]
+pub struct NetLinkJSON {
+    pub id: String,
+    #[serde(rename = "ethernet_mac_address")]
+    pub mac_addr: String,
+}
+
+/// JSON entry in `networks` array.
+#[derive(Debug, Deserialize)]
+pub struct NetNetworkJSON {
+    /// Unique network ID.
+    pub id: String,
+    /// Network type (e.g. `ipv4`)
+    #[serde(rename = "type")]
+    pub kind: String,
+    /// Reference to the underlying interface (see `NetLinkJSON.id`)
+    pub link: String,
+    /// IP network address.
+    pub ip_address: IpAddr,
+    /// IP network mask.
+    pub netmask: IpAddr,
+    /// Routable networks.
+    pub routes: Vec<NetRouteJSON>,
+}
+
+/// JSON entry in `networks.routes` array.
+#[derive(Debug, Deserialize)]
+pub struct NetRouteJSON {
+    /// Route network address.
+    pub network: IpAddr,
+    /// Route netmask.
+    pub netmask: IpAddr,
+    /// Route gateway.
+    pub gateway: IpAddr,
+}
+
+/// JSON entry in `services` array.
+#[derive(Debug, Deserialize)]
+pub struct NetServiceJSON {
+    #[serde(rename = "type")]
+    pub kind: String,
+    pub address: IpAddr,
+}
+
+impl PowerVSProvider {
+    /// Try to build a new provider client.
+    ///
+    /// This internally tries to mount (and own) the config-drive.
+    pub fn try_new() -> Result<Self> {
+        let target = tempfile::Builder::new()
+            .prefix("afterburn-")
+            .tempdir()
+            .context("failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_FS_LABEL),
+            target.path(),
+            CONFIG_DRIVE_FS_TYPE,
+            3, // maximum retries
+        )?;
+
+        let provider = Self {
+            drive_path: target.path().to_owned(),
+            temp_dir: target,
+        };
+        Ok(provider)
+    }
+
+    /// Return the path to the metadata directory.
+    fn metadata_dir(&self) -> PathBuf {
+        let drive = self.drive_path.clone();
+        drive.join("openstack").join("latest")
+    }
+
+    /// Read and parse metadata file.
+    fn read_metadata(&self) -> Result<MetaDataJSON> {
+        let filename = self.metadata_dir().join("meta_data.json");
+        let file = File::open(&filename)
+            .with_context(|| format!("failed to open file '{:?}'", filename))?;
+        let bufrd = BufReader::new(file);
+        Self::parse_metadata(bufrd)
+    }
+
+    /// The public key is stored as key:value pair in openstack/latest/meta_data.json file
+    fn fetch_publickeys(&self) -> Result<Vec<PublicKey>> {
```

Comment:
Hmm, actually, I misled you.  It would be better if this was `fn public_keys(metadata: MetaDataJSON)`, so the tests could verify the output of this function against a fixture.

Reply:
yes..makes sense...changed the test